### ISSUE TITLE
Dot notation (use of the member-access) operator doesn't work with a generic dictionary with [object]-typed keys

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#15997]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Note: This problem won't arise often in real-world usage, given that the very point of using a _generic_ dictionary is to use _specific_ types rather than `[object]`. Still, the behavior is surprising.

With non-generic dictionary types, dot notation can be used as syntactic sugar for index notation; e.g., `@{ foo = 42}.foo` in lieu of `@{ foo = 42 }['foo']`

With a generic dictionary _whose key type is `[object]`_, this doesn't work, however (it does work with `[string]` and `[int]`, for instance, and index notation works in all cases):

```powershell
# Note: $dict['foo'] works fine.
#       $dict.foo would work only if the key type were changed to [string]
(($dict = [System.Collections.Generic.Dictionary[object, object]]::new()).Add('foo', 42)); $dict.foo
```

### Expected behavior

```console
42
```

### Actual behavior

No output.

### Environment data

```powershell
PowerShell Core 7.2.0-preview.9
```

**Comment by daxian-dbw** [comment]
Key with `object` type works with dynamic member access (except when the dynamic member's value is of string type). check this out:
```
$dict = [System.Collections.Generic.Dictionary[object, object]]::new()
$s = [System.IO.FileInfo]::new('bar')
$dict.Add($s, 888)
$dict.$s
> 888
```
when the dynamic member is of the `string` type, it will fall back to the non-dynamic get member binder, where it differentiates `IDictionary` from `IDictionary<,>`. See the code here: https://github.com/PowerShell/PowerShell/blob/6de53f89cf68d957f3c235d847264308ea31ff5a/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L5333-L5354

it's not clear to me why we are dealing with `IDictionary<,>` specially instead of let `IDictionary` to cover it all, just like how dynamic member binder works: https://github.com/PowerShell/PowerShell/blob/6de53f89cf68d957f3c235d847264308ea31ff5a/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L1345-L1356

anyways, when it goes through the `IDictionary<,>` code path, it explicitly handles `IDictionary<string,T>` only.
so in summary, there is inconsistency for sure, as if `$dict.$s` from the above snippet works, I would expect the following works too:
```
$dict.Add('foo', 42)
$v = 'foo'
$dict.$v
```
however, it doesn't

**Comment by vexx32** [comment]
@daxian-dbw likely we need to deal with both because not all `IDictionary<,>` objects also implement `IDictionary`.

**Comment by mklement0** [comment]
Good point, @vexx32 (`ExpandoObject` is an example of such a type - see #15204).

**Comment by daxian-dbw** [comment]
@vexx32 But today we have a check before differentiating `IDictionary<,>` from `IDictionary`:

https://github.com/PowerShell/PowerShell/blob/6de53f89cf68d957f3c235d847264308ea31ff5a/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L5330-L5333

That means we are not dealing with types that doesn't implement `IDictionary` today.

**Comment by daxian-dbw** [comment]
The Engine WG agrees that there is a consistency issue in how `PSGetMemberBinder` and `PSGetDynamicMemberBinder` deal with generic dictionary object. More investigation is needed to understand why `IDictionary<string, T>` is differentiated from `IDictionary` in `PSGetMemberBinder`.
